# Para hoy
1. Especificidad
2. Sensibilidad

Cuando se hace una clasificación en un dataset con desbalanceo de clases se usa el recall o la sensibilidad y la specifity, no la presición. 
si hay balanceo de clases, se usa la precisión 

In [8]:
!pip install sklearn

     |████████████████████████████████| 8.0 MB 3.0 MB/s eta 0:00:01
  Created wheel for sklearn: filename=sklearn-0.0-py2.py3-none-any.whl size=1310 sha256=4f5d21cc53c8d52249e5bbb3ce8556f30dbe0ecbdb05c14e04e5fa4e90ae3105
  Stored in directory: /Users/matiasinostrozacifuentes/Library/Caches/pip/wheels/e4/7b/98/b6466d71b8d738a0c547008b9eb39bf8676d1ff6ca4b22af1c
Successfully built sklearn


In [9]:
import pandas as pd
from numpy.random import randn
from sklearn import metrics

In [10]:
file='/Users/matiasinostrozacifuentes/Desktop/python_intro/DataScience_Unilever/Semana_4/Data/heart_2020_cleaned.csv'
data_corazon=pd.read_csv(file)

In [17]:
data_corazon

HeartDisease    BMI Smoking AlcoholDrinking Stroke  PhysicalHealth  \
0                No  16.60     Yes              No     No             3.0   
1                No  20.34      No              No    Yes             0.0   
2                No  26.58     Yes              No     No            20.0   
3                No  24.21      No              No     No             0.0   
4                No  23.71      No              No     No            28.0   
...             ...    ...     ...             ...    ...             ...   
319790          Yes  27.41     Yes              No     No             7.0   
319791           No  29.84     Yes              No     No             0.0   
319792           No  24.24      No              No     No             0.0   
319793           No  32.81      No              No     No             0.0   
319794           No  46.56      No              No     No             0.0   

        MentalHealth DiffWalking     Sex  AgeCategory      Race Diabetic  \
0               30.0          No  Female        55-59     White      Yes   
1                0.0          No  Female  80 or older     White       No   
2               30.0          No    Male        65-69     White      Yes   
3                0.0          No  Female        75-79     White       No   
4                0.0         Yes  Female        40-44     White       No   
...              ...         ...     ...          ...       ...      ...   
319790           0.0         Yes    Male        60-64  Hispanic      Yes   
319791           0.0          No    Male        35-39  Hispanic       No   
319792           0.0          No  Female        45-49  Hispanic       No   
319793           0.0          No  Female        25-29  Hispanic       No   
319794           0.0          No  Female  80 or older  Hispanic       No   

       PhysicalActivity  GenHealth  SleepTime Asthma KidneyDisease SkinCancer  \
0                   Yes  Very good        5.0    Yes            No        Yes   
1                   Yes  Very good        7.0     No            No         No   
2                   Yes       Fair        8.0    Yes            No         No   
3                    No       Good        6.0     No            No        Yes   
4                   Yes  Very good        8.0     No            No         No   
...                 ...        ...        ...    ...           ...        ...   
319790               No       Fair        6.0    Yes            No         No   
319791              Yes  Very good        5.0    Yes            No         No   
319792              Yes       Good        6.0     No            No         No   
319793               No       Good       12.0     No            No         No   
319794              Yes       Good        8.0     No            No         No   

       prediction  
0             Yes  
1              No  
2             Yes  
3              No  
4             Yes  
...           ...  
319790         No  
319791        Yes  
319792        Yes  
319793        Yes  
319794        Yes  

[319795 rows x 19 columns]

In [12]:
data_corazon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319795 entries, 0 to 319794
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   HeartDisease      319795 non-null  object 
 1   BMI               319795 non-null  float64
 2   Smoking           319795 non-null  object 
 3   AlcoholDrinking   319795 non-null  object 
 4   Stroke            319795 non-null  object 
 5   PhysicalHealth    319795 non-null  float64
 6   MentalHealth      319795 non-null  float64
 7   DiffWalking       319795 non-null  object 
 8   Sex               319795 non-null  object 
 9   AgeCategory       319795 non-null  object 
 10  Race              319795 non-null  object 
 11  Diabetic          319795 non-null  object 
 12  PhysicalActivity  319795 non-null  object 
 13  GenHealth         319795 non-null  object 
 14  SleepTime         319795 non-null  float64
 15  Asthma            319795 non-null  object 
 16  KidneyDisease     31

In [14]:
data_corazon['HeartDisease'].value_counts()

No     292422
Yes     27373
Name: HeartDisease, dtype: int64

In [15]:
Prediction=[]
for i in randn(319795):
    if i <=0:
        Prediction.append('No')
    elif i>0:
        Prediction.append('Yes')
data_corazon['prediction']=Prediction

In [16]:
data_corazon['prediction']

0         Yes
1          No
2         Yes
3          No
4         Yes
         ... 
319790     No
319791    Yes
319792    Yes
319793    Yes
319794    Yes
Name: prediction, Length: 319795, dtype: object

In [18]:
cm = metrics.confusion_matrix(data_corazon['HeartDisease'],data_corazon['prediction'])

In [20]:
cm.ravel()

array([145811, 146611,  13744,  13629])

In [22]:
suma=0
for i in cm.ravel():
    suma+=i
print(suma)

319795


In [23]:
tn, fp, fn, tp = cm.ravel()

In [24]:
print(f'la cantidad de verdaderos positivos es: {tp}\n la cantidad de verdaderos negativos es: {tn}\n la cantidad de falsos positivos es: {fp}\n la cantidad de falsos negativos es: {fn} ')

la cantidad de verdaderos positivos es: 13629
 la cantidad de verdaderos negativos es: 145811
 la cantidad de falsos positivos es: 146611
 la cantidad de falsos negativos es: 13744 


Esto se puede interpretar como:
1. la cantidad de casos que se predijo que tenian problemas al corazon y de verdad lo tenian son 13629.
2. la cantidad de casos que se predijo que no tenian problemas al corazon y no lo tenian son de 145811
3. la cantidad de casos que no tenian y los predijo que sí lo tenian son de 146611
4. la cantidad de casos que tenian y los predijo cono que no lo tenian son de 13744
A primeras impresiones la predicción no es certera, debido a que se generan muchos casos de falsos y los casos de TP son bastante bajos, por eso se procederá a calcular las metricas correspondientes.

In [25]:
sensitivity = tp/ (tp + fn)
specificity = tn / (tn + fp)
precision=tp/(tp+fp)
type_one_error = 1 - specificity
type_two_error = 1 - sensitivity
print('Sensitivity: {:.3f}'.format(sensitivity))
print('Specificity: {:.3f}'.format(specificity))
print('Precisión: {:.3f}'.format(precision))
print('Type 1 Error Rate: {:.3f}'.format(type_one_error))
print('Type 2 Error Rate: {:.3f}'.format(type_two_error))

Sensitivity: 0.498
Specificity: 0.499
Precisión: 0.085
Type 1 Error Rate: 0.501
Type 2 Error Rate: 0.502


Claramente, esta predicción se comporta de manera muy poco precisa, debido a que tanto la sensibilidad como la especificidad es de casí 50%, al igual que las tasas de tipos de errores(I y II), pero hace un claro sentido a como fueron distribuidos los valores de la columna'Prediction' dado a que son numeros que se distribuyen de forma normal. 
Viendo el valor de la precisión vemos que es bastante baja y esto es claro debido a que los valores de casos que de verdad tenian un problema al corazon y los predijo como tal, estaban muy por debajo de el resto de las otras categorias.

---
Se va a hacer un examen
en grupo
Se va a aplicar 
limpieza
preprocesamiento:
-info del data frame
-describe del dataframe
-nulos
-duplicados
-balance de clases
tratamiento:
-como trato los nulos? (eliminar,reemplazar)
-cambiar nombres de columnas de ser necesario
-cambiar valores de datos
-Describe ya limpio 
visualización (minimo 2 gráficos)
y las metricas

In [ ]:
import pandas as pd
import numpy as np

In [26]:
df=pd.read_csv('/Users/matiasinostrozacifuentes/Desktop/python_intro/DataScience_Unilever/Semana_4/Data/StudentsPerformance.csv')

In [27]:
df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

In [28]:
df['test preparation course'].value_counts()

none         642
completed    358
Name: test preparation course, dtype: int64

Primeras conclusiones del preprocesamiento
No hay presencia de nulos
No hay duplicados
No se realizaran reemplazos 
Las columnas son sencillas de llamar y manejar

---

In [ ]:
#Algunos gráficos


---

Ahora se realizará la predicciones de la data, para esto se generará una variable auxiliar con un ciclo for y un random que selecciones valores definidos en una lista y los asigne a la variable auxiliar, para luego agregar esta variable auxiliar al dataframe.

Posterior a esto se realizará una matriz de confusion con sus respectivas metricas

In [29]:
df["Prediction"]=0

import random
lista = ["completed", "none"]
for i in range(0, len(df["Prediction"])):
    x = random.choice(lista)
    df["Prediction"][i]=x

/var/folders/6y/9fksy1pj1bnbf51v6nyzx6fc0000gn/T/ipykernel_2914/1143273437.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Prediction"][i]=x


In [30]:
df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

    Prediction  
0    completed  
1    completed  
2         none  
3    completed  
4         none  
..         ...  
995       none  
996  completed  
997       none  
998  completed  
999       none  

[1000 rows x 9 columns]

In [31]:
len(df['Prediction'])

1000

Creando la matriz de confusión 

In [32]:
from sklearn import metrics #importandola

In [33]:
cm = metrics.confusion_matrix(df['test preparation course'],df['Prediction'])

In [35]:
cm.ravel()

array([186, 172, 297, 345])

In [36]:
tn, fp, fn, tp = cm.ravel()

In [37]:
print(f'la cantidad de verdaderos positivos es: {tp}\n la cantidad de verdaderos negativos es: {tn}\n la cantidad de falsos positivos es: {fp}\n la cantidad de falsos negativos es: {fn} ')

la cantidad de verdaderos positivos es: 345
 la cantidad de verdaderos negativos es: 186
 la cantidad de falsos positivos es: 172
 la cantidad de falsos negativos es: 297 


A primeras impresiones esta clasificando bien, debido a que se ve una concentración mayor a los verdaderos positivos
mientras que en los verdaderons negativos esta en un nivel muy parecido a los falsos positivos.
Esto quiere decir que de la cantidad de estudiantes que no hicieron el curso de preparación, fueron clasificados como tal 345 casos. 

Calculando las metricas

In [38]:
sensitivity = tp/ (tp + fn)
specificity = tn / (tn + fp)
precision=tp/(tp+fp)
type_one_error = 1 - specificity
type_two_error = 1 - sensitivity
print('Sensitivity: {:.3f}'.format(sensitivity))
print('Specificity: {:.3f}'.format(specificity))
print('Precisión: {:.3f}'.format(precision))
print('Type 1 Error Rate: {:.3f}'.format(type_one_error))
print('Type 2 Error Rate: {:.3f}'.format(type_two_error))

Sensitivity: 0.537
Specificity: 0.520
Precisión: 0.667
Type 1 Error Rate: 0.480
Type 2 Error Rate: 0.463


La precisión de este modelo es bastante alto, corroborando lo que se dedujo desde un principio, pero no es menor el nivel de las tasas de los tipos de error, lo que indica que a pesar de ser bastante preciso genera muchos errores de clasificación. Aunque hay que considerar que las categorias que se estan evaluando estan bastante desbalanceadas, por lo que el nivel de precision no es tan certero.

---
ejercicio de los compañeros

In [43]:
df=pd.read_csv('/Users/matiasinostrozacifuentes/Desktop/python_intro/DataScience_Unilever/Semana_4/Data/Life Expectancy Data.csv')

In [40]:
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

In [44]:
df=df.dropna()

In [45]:
df.describe()

Year  Life expectancy   Adult Mortality  infant deaths  \
count  1649.000000       1649.000000      1649.000000    1649.000000   
mean   2007.840509         69.302304       168.215282      32.553062   
std       4.087711          8.796834       125.310417     120.847190   
min    2000.000000         44.000000         1.000000       0.000000   
25%    2005.000000         64.400000        77.000000       1.000000   
50%    2008.000000         71.700000       148.000000       3.000000   
75%    2011.000000         75.000000       227.000000      22.000000   
max    2015.000000         89.000000       723.000000    1600.000000   

           Alcohol  percentage expenditure  Hepatitis B       Measles   \
count  1649.000000             1649.000000  1649.000000    1649.000000   
mean      4.533196              698.973558    79.217708    2224.494239   
std       4.029189             1759.229336    25.604664   10085.802019   
min       0.010000                0.000000     2.000000       0.000000   
25%       0.810000               37.438577    74.000000       0.000000   
50%       3.790000              145.102253    89.000000      15.000000   
75%       7.340000              509.389994    96.000000     373.000000   
max      17.870000            18961.348600    99.000000  131441.000000   

              BMI   under-five deaths         Polio  Total expenditure  \
count  1649.000000         1649.000000  1649.000000        1649.000000   
mean     38.128623           44.220133    83.564585           5.955925   
std      19.754249          162.897999    22.450557           2.299385   
min       2.000000            0.000000     3.000000           0.740000   
25%      19.500000            1.000000    81.000000           4.410000   
50%      43.700000            4.000000    93.000000           5.840000   
75%      55.800000           29.000000    97.000000           7.470000   
max      77.100000         2100.000000    99.000000          14.390000   

       Diphtheria      HIV/AIDS            GDP    Population  \
count  1649.000000  1649.000000    1649.000000  1.649000e+03   
mean     84.155246     1.983869    5566.031887  1.465363e+07   
std      21.579193     6.032360   11475.900117  7.046039e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      82.000000     0.100000     462.149650  1.918970e+05   
50%      92.000000     0.100000    1592.572182  1.419631e+06   
75%      97.000000     0.700000    4718.512910  7.658972e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

        thinness  1-19 years   thinness 5-9 years  \
count            1649.000000          1649.000000   
mean                4.850637             4.907762   
std                 4.599228             4.653757   
min                 0.100000             0.100000   
25%                 1.600000             1.700000   
50%                 3.000000             3.200000   
75%                 7.100000             7.100000   
max                27.200000            28.200000   

       Income composition of resources    Schooling  
count                      1649.000000  1649.000000  
mean                          0.631551    12.119891  
std                           0.183089     2.795388  
min                           0.000000     4.200000  
25%                           0.509000    10.300000  
50%                           0.673000    12.300000  
75%                           0.751000    14.000000  
max                           0.936000    20.700000

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1649 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          1649 non-null   object 
 1   Year                             1649 non-null   int64  
 2   Status                           1649 non-null   object 
 3   Life expectancy                  1649 non-null   float64
 4   Adult Mortality                  1649 non-null   float64
 5   infant deaths                    1649 non-null   int64  
 6   Alcohol                          1649 non-null   float64
 7   percentage expenditure           1649 non-null   float64
 8   Hepatitis B                      1649 non-null   float64
 9   Measles                          1649 non-null   int64  
 10   BMI                             1649 non-null   float64
 11  under-five deaths                1649 non-null   int64  
 12  Polio               

In [48]:
df['Status'].value_counts()

Developing    1407
Developed      242
Name: Status, dtype: int64

Data no esta balanceada! no sería útil ocupar la metrica de precisión

In [51]:
c=[]
import random
lista = ["Developing", "Developed"]
for i in range(0, len(df["Prediction"])):
    x = random.choice(lista)
    c.append(x)
df['Prediction']=c

In [52]:
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0     ...               8.16         65.0        0.1  584.259210  33736494.0   
1     ...               8.18         62.0        0.1  612.696514    327582.0   
2     ...               8.13         64.0        0.1  631.744976  31731688.0   
3     ...               8.52         67.0        0.1  669.959000   3696958.0   
4     ...               7.87         68.0        0.1   63.537231   2978599.0   
...   ...                ...          ...        ...         ...         ...   
2933  ...               7.13         65.0       33.6  454.366654  12777511.0   
2934  ...               6.52         68.0       36.7  453.351155  12633897.0   
2935  ...               6.53         71.0       39.8   57.348340    125525.0   
2936  ...               6.16         75.0       42.1  548.587312  12366165.0   
2937  ...               7.10         78.0       43.5  547.358878  12222251.0   

       thinness  1-19 years   thinness 5-9 years  \
0                      17.2                 17.3   
1                      17.5                 17.5   
2                      17.7                 17.7   
3                      17.9                 18.0   
4                      18.2                 18.2   
...                     ...                  ...   
2933                    9.4                  9.4   
2934                    9.8                  9.9   
2935                    1.2                  1.3   
2936                    1.6                  1.7   
2937                   11.0                 11.2   

      Income composition of resources  Schooling  Prediction  
0                               0.479       10.1   Developed  
1                               0.476       10.0  Developing  
2                               0.470        9.9   Developed  
3                               0.463        9.8  Developing  
4                               0.454        9.5  Developing  
...                               ...        ...         ...  
2933                            0.407        9.2  Developing  
2934                            0.418        9.5   Developed  
2935               

In [53]:
cm = metrics.confusion_matrix(df['Status'],df['Prediction'])

In [54]:
cm.ravel()

array([125, 117, 692, 715])

In [55]:
tn, fp, fn, tp = cm.ravel()

In [57]:
print(f'la cantidad de verdaderos positivos es: {tp}\n la cantidad de verdaderos negativos es: {tn}\n la cantidad de falsos positivos es: {fp}\n la cantidad de falsos negativos es: {fn} ')

la cantidad de verdaderos positivos es: 715
 la cantidad de verdaderos negativos es: 125
 la cantidad de falsos positivos es: 117
 la cantidad de falsos negativos es: 692 


En este caso, la cantidad de verdaderos positivos es bastante alta, al igual que la de falsos negativos lo que da indicios que el modelo no ha de ser muy correcto, si revisamos las metricas:

In [58]:
sensitivity = tp/ (tp + fn)
specificity = tn / (tn + fp)
precision=tp/(tp+fp)
type_one_error = 1 - specificity
type_two_error = 1 - sensitivity
print('Sensitivity: {:.3f}'.format(sensitivity))
print('Specificity: {:.3f}'.format(specificity))
print('Precisión: {:.3f}'.format(precision))
print('Type 1 Error Rate: {:.3f}'.format(type_one_error))
print('Type 2 Error Rate: {:.3f}'.format(type_two_error))

Sensitivity: 0.508
Specificity: 0.517
Precisión: 0.859
Type 1 Error Rate: 0.483
Type 2 Error Rate: 0.492


Mientras que claro, los niveles de sensibilidad y de especificidad son bastante parecidos, al igual que las tasas de error, mostrando que a pesar que la precision es muy alta, los niveles de las metricas no son los ideales y permite concluir que la predicción puede ser precisa, pero no es exacta a los valores que se tiene.